runtime/near-vm-logic/src/gas_counter.rs

```diff
@@ -49,48 +69,94 @@ impl GasCounter {
     pub fn new(
         ext_costs_config: ExtCostsConfig,
         max_gas_burnt: Gas,
+        opcode_cost: u32,
         prepaid_gas: Gas,
         is_view: bool,
     ) -> Self {
+        use std::cmp::min;
         Self {
             ext_costs_config,
-            burnt_gas: 0,
-            used_gas: 0,
-            max_gas_burnt,
+            fast_counter: FastGasCounter {
+                burnt_gas: 0,
+                gas_limit: min(max_gas_burnt, prepaid_gas),
+                opcode_cost: Gas::from(opcode_cost),
+            },
+            max_gas_burnt: max_gas_burnt,
+            promises_gas: 0,
             prepaid_gas,
             is_view,
             profile: Default::default(),
         }
     }
 
     fn deduct_gas(&mut self, burn_gas: Gas, use_gas: Gas) -> Result<()> {
+        use std::cmp::min;
         assert!(burn_gas <= use_gas);
+        let promise_gas = use_gas - burn_gas;
+        let new_promises_gas =
+            self.promises_gas.checked_add(promise_gas).ok_or(HostError::IntegerOverflow)?;
         let new_burnt_gas =
-            self.burnt_gas.checked_add(burn_gas).ok_or(HostError::IntegerOverflow)?;
-        let new_used_gas = self.used_gas.checked_add(use_gas).ok_or(HostError::IntegerOverflow)?;
+            self.fast_counter.burnt_gas.checked_add(burn_gas).ok_or(HostError::IntegerOverflow)?;
+        let new_used_gas =
+            new_burnt_gas.checked_add(new_promises_gas).ok_or(HostError::IntegerOverflow)?;
         if new_burnt_gas <= self.max_gas_burnt && (self.is_view || new_used_gas <= self.prepaid_gas)
         {
-            self.burnt_gas = new_burnt_gas;
-            self.used_gas = new_used_gas;
+            if promise_gas != 0 {
+                self.fast_counter.gas_limit =
+                    min(self.max_gas_burnt, self.prepaid_gas - new_promises_gas);
+            }
+            self.fast_counter.burnt_gas = new_burnt_gas;
+            self.promises_gas = new_promises_gas;
             Ok(())
         } else {
-            use std::cmp::min;
-            let res = if new_burnt_gas > self.max_gas_burnt {
+            // if max_gas_burnt == prepaid_gas we must use GasExceeded error.
+            if new_burnt_gas > self.max_gas_burnt && self.max_gas_burnt != self.prepaid_gas {
+                self.fast_counter.burnt_gas = self.max_gas_burnt;
+                self.promises_gas = 0;
```

Comment:
If I comment-out this line, the tests seem to pass. That's suspicious. Can we add a test that shows that this line is needed?

Reply:
We can add this line when making protocol change.